src/core/experimental/array.d

```diff
@@ -0,0 +1,2559 @@
+        static assert(!__traits(compiles, () @safe { Array!Unsafe(Unsafe(1)); }));
+        static assert(!__traits(compiles, () @safe { auto a = const Array!Unsafe(Unsafe(1)); }));
+        static assert(!__traits(compiles, () @safe { auto a = immutable Array!Unsafe(Unsafe(1)); }));
+
+        static assert(!__traits(compiles, () @safe { Array!UnsafeDtor(UnsafeDtor(1)); }));
+        static assert(!__traits(compiles, () @safe { auto s = const Array!UnsafeDtor(UnsafeDtor(1)); }));
+        static assert(!__traits(compiles, () @safe { auto s = immutable Array!UnsafeDtor(UnsafeDtor(1)); }));
+
+        // Infer purity
+        static assert(!__traits(compiles, () pure { Array!Impure(Impure(1)); }));
+        static assert(!__traits(compiles, () pure { auto a = const Array!Impure(Impure(1)); }));
+        static assert(!__traits(compiles, () pure { auto a = immutable Array!Impure(Impure(1)); }));
+
+        static assert(!__traits(compiles, () pure { Array!ImpureDtor(ImpureDtor(1)); }));
+        static assert(!__traits(compiles, () pure { auto s = const Array!ImpureDtor(ImpureDtor(1)); }));
+        static assert(!__traits(compiles, () pure { auto s = immutable Array!ImpureDtor(ImpureDtor(1)); }));
+
+        // Infer throwability
+        static assert(!__traits(compiles, () nothrow { Array!Throws(Throws(1)); }));
+        static assert(!__traits(compiles, () nothrow { auto a = const Array!Throws(Throws(1)); }));
+        static assert(!__traits(compiles, () nothrow { auto a = immutable Array!Throws(Throws(1)); }));
+
+        static assert(!__traits(compiles, () nothrow { Array!ThrowsDtor(ThrowsDtor(1)); }));
+        static assert(!__traits(compiles, () nothrow { auto s = const Array!ThrowsDtor(ThrowsDtor(1)); }));
+        static assert(!__traits(compiles, () nothrow { auto s = immutable Array!ThrowsDtor(ThrowsDtor(1)); }));
+    }
+
+    private @nogc nothrow pure @trusted
+    size_t slackFront() const
+    {
+        return _payload.ptr - _support.ptr;
+    }
+
+    private @nogc nothrow pure @trusted
+    size_t slackBack() const
+    {
+        return _support.ptr + _support.length - _payload.ptr - _payload.length;
+    }
+
+    /**
+     * Return the number of elements in the array..
+     *
+     * Returns:
+     *      the length of the array.
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    @nogc nothrow pure @safe
+    size_t length() const
+    {
+        return _payload.length;
+    }
+
+    /// ditto
+    alias opDollar = length;
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto a = Array!int(1, 2, 3);
+        assert(a.length == 3);
+        assert(a[$ - 1] == 3);
+    }
+
+    /**
+     * Set the length of the array to `len`. `len` must be less than or equal
+     * to the `capacity` of the array.
+     *
+     * Params:
+     *      len = a positive integer
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    @nogc nothrow pure @trusted
+    void forceLength(size_t len)
```

Comment:
Why isn’t this called `length`?

Reply:
It was't doing exactly as built-in array `length`. I'll update the PR with the updated name and implementation.